animations-dbus/animations-dbus-server-object.c

```diff
@@ -735,3 +744,68 @@ animations_dbus_server_new_with_connection_async (AnimationsDbusServerEffectFact
                               "effect-factory", factory,
                               NULL);
 }
+
+/**
+ * animations_dbus_server_stop:
+ * @self: an #AnimationsDbusServer
+ * @cancellable: (nullable): a #GCancellable, or %NULL
+ * @error: return location for a #GError, or %NULL
+ *
+ * Stop the server, unexport all D-Bus objects, and unown the well-known
+ * D-Bus name for the service.
+ *
+ * Returns: %TRUE on success, %FALSE otherwise
+ */
+gboolean
+animations_dbus_server_stop (AnimationsDbusServer  *self,
+                             GCancellable          *cancellable,
+                             GError               **error)
+{
+  AnimationsDbusServerPrivate *priv = animations_dbus_server_get_instance_private (self);
+  guint i;
+  GHashTableIter iter;
+  gboolean error_seen = FALSE;
+
+  g_return_val_if_fail (ANIMATIONS_DBUS_IS_SERVER (self), FALSE);
+  g_return_val_if_fail (cancellable == NULL || G_IS_CANCELLABLE (cancellable), FALSE);
+  g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
+
+  for (i = 0; priv->animatable_surfaces != NULL && i < priv->animatable_surfaces->len; i++)
+    {
+      AnimationsDbusServerSurface *surface = g_ptr_array_index (priv->animatable_surfaces, i);
+
+      if (!animations_dbus_server_unregister_surface (self, surface, NULL))
+        error_seen = TRUE;
```

Comment:
Here you are iterating while removing from the same array, isn't that wrong given `array->len` is decreasing while the index (`i`) is increasing at the same time? I think a while loop would probably be clearer:
```
while (priv->animatable_surfaces->len > 0)
  {
     AnimationsDbusServerSurface *surface = g_ptr_array_index (priv->animatable_surfaces, 0);

      if (!animations_dbus_server_unregister_surface (self, surface, NULL))
        error_seen = TRUE;
  }
```

Reply:
Good catch, fixed.